project/DerivedJoiningTypeCodeGen.scala

```diff
@@ -0,0 +1,293 @@
+   * Unicode version.
+   */
+  def rowsFromUnicodeVersion(
+      unicodeVersion: UnicodeVersion): Either[Throwable, SortedSet[Row]] =
+    Either
+      .catchNonFatal(
+        new URL(
+          s"https://www.unicode.org/Public/${unicodeVersion.asString}/ucd/extracted/DerivedJoiningType.txt")
+      )
+      .flatMap(url => rowsFromUrl(url))
+
+  /**
+   * Attempt to convert some [[Row]] values into a mapping between [[CodePointRange]] values and
+   * [[JoiningType]] values.
+   */
+  def joiningTypeMap[F[_]: Foldable](
+      rows: F[Row]): Either[String, SortedMap[CodePointRange, JoiningType]] =
+    flattenValues(
+      rows.foldMap {
+        case Row(codePointRange, joiningType) =>
+          SortedMap(codePointRange -> NonEmptyChain.one(joiningType))
+      }
+    )
+
+  /**
+   * Create the source code for the generated file.
+   */
+  def generateSource(value: SortedMap[CodePointRange, JoiningType]): Tree =
+    source"""
+package org.typelevel.idna4s.core.uts46
+
+import scala.collection.immutable.IntMap
+
+private[uts46] trait ${Type.Name(GeneratedTypeName)} extends ${Init(
+        Type.Name(BaseTypeName),
+        scala.meta.Name(""),
+        Nil)} {
+..${joiningTypeMapDef(value)}
+}
+"""
+
+  /**
+   * Create the Defn values for the [[CodePointRange]] to [[JoiningType]] mappings. These Defn
+   * values will implement the interface defined in `JoiningTypeBase`.
+   */
+  def joiningTypeMapDef(value: SortedMap[CodePointRange, JoiningType]): List[Defn] = {
+    val (singles, ranges): (
+        SortedMap[CodePointRange.Single, JoiningType],
+        SortedMap[CodePointRange, JoiningType]) =
+      value
+        .toVector
+        .foldLeft(
+          (
+            SortedMap.empty[CodePointRange.Single, JoiningType],
+            SortedMap.empty[CodePointRange, JoiningType])) {
+          case ((singles, ranges), (cp: CodePointRange.Single, joiningType)) =>
+            (singles + (cp -> joiningType), ranges)
+          case ((singles, ranges), (cp, joiningType)) =>
+            (singles, ranges + (cp -> joiningType))
+        }
+
+    val rangeTerms: List[Term] =
+      ranges.toList.map {
+        case (k, v) =>
+          q"(Range.inclusive(${Lit.Int(k.lower.value)}, ${Lit.Int(k.upper.value)}), ${Lit.Char(v.value)})"
+      }
+
+    val singleTerms: List[Term] =
+      singles.toList.map {
+        case (k, v) =>
+          q"(${Lit.Int(k.value.value)}, ${Lit.Char(v.value)})"
+      }
+
+    def baseMap: Term =
+      q"IntMap(..$singleTerms)"
+
+    val rhs: Term =
+      if (singleTerms.isEmpty && rangeTerms.isEmpty) {
+        q"IntMap.empty[Char]"
+      } else if (rangeTerms.isEmpty) {
+        baseMap
+      } else {
+        q"List(..$rangeTerms).foldLeft($baseMap){case (acc, (range, value)) => range.foldLeft(acc){case (acc, cp) => acc.updated(cp, value)}}"
+      }
+
+    List(
+      q"""override final protected lazy val joiningTypeMap: IntMap[Char] = $rhs"""
```

Comment:
Tangential micro-optimization: `lazy val`s have various synchronization costs that apply to every access. If we care enough we may want to use a manual, unsafe encoding i.e. `def` + `var`. There is a risk it may init more than once, but only one will "win" post-warmup and can be accessed without any synchronization.

https://dotty.epfl.ch/docs/reference/other-new-features/threadUnsafe-annotation.html

Reply:
@armanbilge so I actually prototyped a few various encodings of using def + var with code point mapper and found that it dramatically slowed down the benchmark. I wanted to go back and take a second look, but it appeared to be related to extra method calls in the bytecode to access the underlying field that were always incurred even after initialization.

I did a little digging, and I noticed something surprising in the bytecode for the current `lazy val joiningTypeMap` encoding.

```
  public final scala.collection.immutable.IntMap<java.lang.Object> joiningTypeMap();
    descriptor: ()Lscala/collection/immutable/IntMap;
    flags: (0x0011) ACC_PUBLIC, ACC_FINAL
    Code:
      stack=2, locals=1, args_size=1
         0: getstatic     #80                 // Field bitmap$0:B
         3: iconst_1
         4: iand
         5: i2b
         6: iconst_0
         7: if_icmpne     15
        10: aload_0
        11: invokespecial #90                 // Method joiningTypeMap$lzycompute:()Lscala/collection/immutable/IntMap;
        14: areturn
        15: getstatic     #86                 // Field joiningTypeMap:Lscala/collection/immutable/IntMap;
        18: areturn
      StackMapTable: number_of_entries = 1
        frame_type = 15 /* same */
      LineNumberTable:
        line 24: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      19     0  this   Lorg/typelevel/idna4s/core/uts46/UTS46$;
    Signature: #78                          // ()Lscala/collection/immutable/IntMap<Ljava/lang/Object;>;
```

If I'm reading that correctly, then if the bitmap$0 field is already set, we avoid synchronization on the monitor lock and just directly return the static field. The monitor lock is acquired in `joiningTypeMap$lzycompute`. If that's true, then this is already basically the same as writing a def + var encoding, right? The only difference being that if more than one thread enters before the static field is set, we will block on thread on the monitor lock, but after the bitmap$0 field is set we bypass all of that.

Let me know if any of that seems wrong to you. Like I said, I wanted to take a second look after all the functionality was in.

Reply:
Ah, interesting, thanks for sharing!

---

> I wanted to go back and take a second look, but it appeared to be related to extra method calls in the bytecode to access the underlying field that were always incurred even after initialization.

Hmm. Did you mark it as a `private[this] var`? You can see the bytecode difference demoed here:
- https://github.com/typelevel/cats/pull/4314#discussion_r988196572

---

> If I'm reading that correctly, then if the bitmap$0 field is already set, we avoid synchronization on the monitor lock and just directly return the static field.

But the `bitmap$0` field is itself volatile, which is a form of synchronization. It's forcing an unnecessary read barrier.

Reply:
I don't recall if I used `private[this]` or not, but I suspect I did not because I wasn't aware it had that type of effect on the bytecode. 

I'll add that to #21 and do a comprehensive pass before the next release. For now, I'd like to leave the `lazy val`s in place just for consistency and then change them all at once if we can show it is better for initialization without negatively affecting the overall performance.

Reply:
Sorry, I'll add this to #20, not #21. :facepalm: